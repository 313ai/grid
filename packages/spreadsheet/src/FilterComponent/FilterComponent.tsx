import React, { useState, useCallback } from "react";
import {
  Box,
  Menu,
  MenuList,
  MenuDivider,
  useColorMode,
  useTheme,
  InputGroup,
  InputLeftElement,
  Icon,
  Input,
  Checkbox,
} from "@chakra-ui/core";
import { DARK_MODE_COLOR } from "../constants";
import { MenuItem, Button } from "./../styled";
import { CellPosition, FilterDefinition } from "@rowsncolumns/grid";

export interface FilterComponentProps {
  position: CellPosition;
  onChange?: (
    filterIndex: number,
    columnIndex: number,
    filter: FilterDefinition
  ) => void;
  onCancel?: () => void;
  values: React.ReactText[];
  filter: FilterDefinition;
  columnIndex?: number;
  index: number;
  width: number;
}

const FilterComponent = ({
  position,
  onChange,
  onCancel,
  values = [],
  filter,
  columnIndex,
  index,
  width,
}: FilterComponentProps) => {
  if (!columnIndex) return null;
  const [filterText, setFilterText] = useState("");
  const { values: selectedValues = [] } = filter || {};
  const [userValues, setUserValues] = useState(
    filter ? selectedValues : values
  );
  const theme = useTheme();
  const { colorMode } = useColorMode();
  const isLight = colorMode === "light";
  const color = isLight ? DARK_MODE_COLOR : theme.colors.white;
  const borderColor = isLight ? theme.colors.gray[300] : DARK_MODE_COLOR;
  const bgColor = isLight ? theme.colors.white : theme.colors.gray[700];
  let { x = 0, y } = position;
  const handleChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    const { value, checked } = e.target;
    if (checked) {
      setUserValues((prev) => prev.concat(value));
    } else {
      setUserValues((prev) => prev.filter((v) => v !== value));
    }
  }, []);
  const handleSubmit = useCallback(() => {
    onChange?.(index, columnIndex, {
      values: userValues,
      operator: filter?.operator,
    });
  }, [userValues]);
  const isSelectAll = values.length === userValues.length;
  const isIndeterminate = !isSelectAll && userValues.length > 0;
  return (
    <Box
      left={0}
      top={0}
      position="absolute"
      zIndex={1}
      transform={`translate(${x}px, ${y}px)`}
    >
      <Box shadow="md" bg={bgColor} width={width}>
        <Box padding={3}>
          <Button
            onClick={() => setUserValues(values)}
            variant="link"
            fontSize={12}
            mr={2}
            background="none"
            size="sm"
          >
            Select all
          </Button>
          <Button
            onClick={() => setUserValues([])}
            variant="link"
            fontSize={12}
            background="none"
            size="sm"
          >
            Clear
          </Button>
          <InputGroup size="sm" mt={2}>
            <InputLeftElement
              children={<Icon name="search" color="gray.300" />}
            />
            <Input
              type="text"
              size="sm"
              borderColor={borderColor}
              value={filterText}
              onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
                setFilterText(e.target.value);
              }}
            />
          </InputGroup>
          <Box overflow="auto" maxHeight={100} mt={2} mb={2} pl={1} pr={1}>
            {values
              .filter((value) =>
                new RegExp(filterText, "gi").test(value.toString())
              )
              .map((value, idx) => {
                const valueItem = value || "";
                const isChecked = userValues.indexOf(valueItem) !== -1;
                const label = value || "(Blanks)";
                return (
                  <Box fontSize={12} key={idx}>
                    <Checkbox
                      onChange={handleChange}
                      value={valueItem}
                      size="sm"
                      borderColor={borderColor}
                      isChecked={isChecked}
                    >
                      {label}
                    </Checkbox>
                  </Box>
                );
              })}
          </Box>
          <Box pt={2}>
            <Button
              size="sm"
              variantColor="green"
              variant="solid"
              onClick={handleSubmit}
              mr={1}
            >
              Done
            </Button>
            <Button size="sm" variant="outline" onClick={onCancel}>
              Cancel
            </Button>
          </Box>
        </Box>
      </Box>
    </Box>
  );
};

export default FilterComponent;
