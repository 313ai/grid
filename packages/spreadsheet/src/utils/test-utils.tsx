import React from "react";
import { render } from "@testing-library/react";
import { Stage, Layer } from "react-konva";
import { ThemeProvider, theme } from "@chakra-ui/core";
import Konva from "konva";

const KonvaWrapper = ({ children }) => {
  return (
    <Stage>
      <ThemeWrapper>
        <Layer>{children}</Layer>
      </ThemeWrapper>
    </Stage>
  );
};

const konvaRenderer = (ui) => render(ui, { wrapper: KonvaWrapper });
const domRenderer = (ui) => render(ui, { wrapper: ThemeWrapper });

export * from "@testing-library/react";
export { konvaRenderer, domRenderer };

export const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
