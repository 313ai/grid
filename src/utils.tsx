import React from "react";
import { Line, Rect } from "react-konva";
import { ShapeConfig } from "konva/types/Shape";

interface BoxProps extends ShapeConfig {
  onCrosshairMouseDown?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseMove?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onCrosshairMouseUp?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Create a box with custom top/right/bottom/left colors and widths
 * @param param0
 */
export const createBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
}: BoxProps) => {
  const commonProps = {
    perfectDrawEnabled: false,
    shadowForStrokeEnabled: false,
    hitStrokeWidth: 0,
    listening: false,
    lineCap: "square",
  };
  const lines = [
    <Line
      points={[x, y, x + width, y]}
      stroke={strokeTopColor}
      strokeWidth={strokeTopWidth}
      key="top"
      {...commonProps}
    />,
    <Line
      points={[x + width, y, x + width, y + height]}
      stroke={strokeRightColor}
      strokeWidth={strokeRightWidth}
      key="right"
      {...commonProps}
    />,
    <Line
      points={[x + width, y + height, x, y + height]}
      stroke={strokeBottomColor}
      strokeWidth={strokeBottomWidth}
      key="bottom"
      {...commonProps}
    />,
    <Line
      points={[x, y + height, x, y]}
      stroke={strokeLeftColor}
      strokeWidth={strokeLeftWidth}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <Rect
          fill={fill}
          x={x}
          y={y}
          width={width}
          height={height}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};

export const FillHandle: React.FC<BoxProps> = ({
  x = 0,
  y = 0,
  stroke,
  onMouseDown,
}) => {
  if (x === 0 || y === 0) return null;
  const crossHairSize = 8;
  const crossHairBorderSize = 1;
  return (
    <div
      style={{
        position: "absolute",
        left: x - crossHairSize / 2,
        top: y - crossHairSize / 2,
        width: crossHairSize,
        height: crossHairSize,
        border: `${crossHairBorderSize}px white solid`,
        background: stroke,
        cursor: "crosshair",
      }}
      onMouseDown={onMouseDown}
    />
  );
};

export const createHTMLBox = ({
  x = 0,
  y = 0,
  width = 0,
  height = 0,
  fill,
  stroke,
  strokeLeftColor = stroke,
  strokeTopColor = stroke,
  strokeRightColor = stroke,
  strokeBottomColor = stroke,
  strokeWidth = 0,
  strokeTopWidth = strokeWidth,
  strokeRightWidth = strokeWidth,
  strokeBottomWidth = strokeWidth,
  strokeLeftWidth = strokeWidth,
  key,
  strokeStyle = "solid",
}: BoxProps) => {
  const commonProps = {};
  width = width - Math.floor(strokeWidth);
  y = y - Math.ceil(strokeWidth / 2);
  const lines = [
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: width,
        height: strokeTopWidth,
        borderWidth: 0,
        borderColor: strokeTopColor,
        borderTopWidth: strokeTopWidth,
        borderStyle: strokeStyle,
      }}
      key="top"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x + width,
        top: y,
        width: strokeRightWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeRightColor,
        borderRightWidth: strokeRightWidth,
        borderStyle: strokeStyle,
      }}
      key="right"
      {...commonProps}
    />,
    ,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y + height,
        width: width + strokeTopWidth,
        height: strokeBottomWidth,
        borderWidth: 0,
        borderColor: strokeBottomColor,
        borderBottomWidth: strokeBottomWidth,
        borderStyle: strokeStyle,
      }}
      key="bottom"
      {...commonProps}
    />,
    <div
      style={{
        position: "absolute",
        left: x,
        top: y,
        width: strokeLeftWidth,
        height: height,
        borderWidth: 0,
        borderColor: strokeLeftColor,
        borderLeftWidth: strokeLeftWidth,
        borderStyle: strokeStyle,
      }}
      key="left"
      {...commonProps}
    />,
  ];

  return (
    <React.Fragment key={key}>
      {lines}
      {fill && (
        <div
          style={{
            position: "absolute",
            top: y,
            left: x,
            height,
            width,
            backgroundColor: fill,
            userSelect: "none",
            pointerEvents: "none",
          }}
          {...commonProps}
        />
      )}
    </React.Fragment>
  );
};
