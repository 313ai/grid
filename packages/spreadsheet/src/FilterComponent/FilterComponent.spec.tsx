import React from "react";
import { domRenderer, cleanup } from "./../utils/test-utils";
import FilterComponent from "./FilterComponent";

describe("FilterComponent", () => {
  afterEach(cleanup);
  it("renders a filterComponent", () => {
    const renderGrid = () =>
      domRenderer(
        <FilterComponent
          columnIndex={1}
          position={{
            x: 0,
            y: 0,
            width: 300,
            height: 500
          }}
          values={
            ['A', 'b']
          }
          filter={{
            operator: 'containsText',
            values: ['H']
          }}
          index={0}
          width={200}
        />
      );
    expect(renderGrid).not.toThrow();
  });

  it("matches snapshot", () => {
    const { asFragment } = domRenderer(
      <FilterComponent
        columnIndex={1}
        position={{
          x: 0,
          y: 0,
          width: 300,
          height: 500
        }}
        values={
          ['A', 'b']
        }
        filter={{
          operator: 'containsText',
          values: ['H']
        }}
        index={0}
        width={200}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
})