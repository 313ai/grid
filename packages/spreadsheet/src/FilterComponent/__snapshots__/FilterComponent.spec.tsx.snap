// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterComponent matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1jm5quw"
  >
    <div
      class="css-hw9y9f"
    >
      <div
        class="css-158s385"
      >
        <button
          class="css-1gopdl6-Button"
          type="button"
        >
          Select all
        </button>
        <button
          class="css-sq4f40-Button"
          type="button"
        >
          Clear
        </button>
        <div
          class="css-770flb"
        >
          <div
            class="css-btmwlx"
          >
            <svg
              class="css-1xb1rov"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
                fill="currentColor"
              />
            </svg>
          </div>
          <input
            class="css-15mnt7s"
            type="text"
            value=""
          />
        </div>
        <div
          class="css-1f3tcdl"
        >
          <div
            style="position: relative;"
          >
            <div
              style="transform: translate(0px, 0px);"
            >
              <div>
                <div
                  class="css-125uqq"
                >
                  <label
                    class="css-18lkzlc"
                  >
                    <input
                      aria-checked="false"
                      class="css-1gy22ay"
                      type="checkbox"
                      value="b"
                    />
                    <div
                      aria-hidden="true"
                      class="css-1d4ijjx"
                    >
                      <svg
                        class="css-1vhqk79"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 14 14"
                      >
                        <g
                          fill="currentColor"
                        >
                          <polygon
                            points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
                          />
                        </g>
                      </svg>
                    </div>
                    <div
                      class="css-75rnni"
                    >
                      <div
                        class="css-1qxtz39"
                      >
                        b
                      </div>
                    </div>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-1p885yg"
        >
          <button
            class="css-x8sq2n-Button"
            type="button"
          >
            Done
          </button>
          <button
            class="css-wighrs-Button"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
